source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.h

```diff
@@ -164,6 +164,30 @@ class LocalRateLimiterImpl : public Logger::Loggable<Logger::Id::local_rate_limi
   const bool always_consume_default_token_bucket_{};
 };
 
+class LocalRateLimiterMapSingleton;
+using LocalRateLimiterMapSingletonSharedPtr = std::shared_ptr<LocalRateLimiterMapSingleton>;
+class LocalRateLimiterMapSingleton : public Singleton::Instance {
+public:
+  struct RateLimiter {
+    // The rate limiter map singleton isn't pinned in the process and is shared among all the
+    // access log rate limit filters, so this struct holds its shared pointer for ownership.
+    LocalRateLimiterMapSingletonSharedPtr map_;
+    LocalRateLimiterImpl& limiter_;
+  };
+
+  static RateLimiter getRateLimiter(
+      Singleton::Manager& manager, absl::string_view limiter_key,
+      const std::chrono::milliseconds fill_interval, const uint64_t max_tokens,
+      const uint64_t tokens_per_fill, Event::Dispatcher& dispatcher,
+      const Protobuf::RepeatedPtrField<
+          envoy::extensions::common::ratelimit::v3::LocalRateLimitDescriptor>& descriptors,
+      bool always_consume_default_token_bucket, ShareProviderSharedPtr shared_provider,
+      const uint32_t lru_size);
+
+private:
+  absl::flat_hash_map<std::string, std::unique_ptr<LocalRateLimiterImpl>> limiter_map_;
```

Comment:
Should it be a weak_ptr in the singleton map? Otherwise, when some filters are done with it, but some other filters use another key, the first one would stay alive.